driver-ui/src/main/java/eu/tsystems/mms/tic/testframework/utils/UITestUtils.java

```diff
@@ -308,32 +308,34 @@ public static List<Screenshot> takeScreenshots() {
     }
 
     /**
-     * Make screenshots from all open browser windows/selenium, webdriver instances of the current thread
-     * and of exclusive webdriver instances of the current method context (if exists).
+     * Make screenshots from all open Webdriver instances the current method context.
      *
      * @param publishToReport True for publish directly into report.
      * @return ScreenshotPaths.
      */
     public static List<Screenshot> takeScreenshots(final boolean publishToReport) {
-        Optional<MethodContext> optionalMethodContext = executionContextController.getCurrentMethodContext();
-
-        // Find exclusive sessions linked to current method context
-        Stream<String> exclusiveSessionKeys = Stream.empty();
-        if (optionalMethodContext.isPresent()) {
-            exclusiveSessionKeys = optionalMethodContext.get().readSessionContexts()
-                    .map(SessionContext::getSessionKey)
-                    .filter(sessionKey -> sessionKey.startsWith(SessionContext.EXCLUSIVE_PREFIX));
+        Optional<MethodContext> methodContext = executionContextController.getCurrentMethodContext();
+
+        if (methodContext.isEmpty()) {
+            LOGGER.warn("Please use this method only in test or setup methods. Otherwise no screenshots are created.");
+            return Collections.emptyList();
         }
 
-        // Take all normal webdriver and linked exclusive webdriver and create screenshots
-        Stream<Screenshot> screenshotStream = Stream.concat(WEB_DRIVER_MANAGER.readWebDriversFromCurrentThread(), exclusiveSessionKeys.map(WEB_DRIVER_MANAGER::getWebDriver))
+        // Find all sessions of current method context and create screenshots
+        Stream<Screenshot> screenshotStream = methodContext.get().readSessionContexts()
+                .map(SessionContext::getSessionKey)
+                .map(WEB_DRIVER_MANAGER::getWebDriver)
                 .map(UITestUtils::pTakeAllScreenshotsForSession)
                 .flatMap(Collection::stream);
 
-        if (publishToReport && optionalMethodContext.isPresent()) {
-            MethodContext methodContext = optionalMethodContext.get();
+        if (screenshotStream.findFirst().isEmpty()) {
```

Comment:
Hold on. `findFirst` is a terminal operation. This will make the Stream unusable further down.
... This branch was not here in the last review, thence I suggested using the Stream directly - but if we do want to use the data more than once, we cannot use Stream and have to use a proper Collection after all.

Reply:
Suggested fix:
```java
// Find all sessions of current method context and create screenshots
Stream<Screenshot> screenshotStream = methodContext.get().readSessionContexts()
		.map(SessionContext::getSessionKey)
		.map(WEB_DRIVER_MANAGER::getWebDriver)
		.map(UITestUtils::pTakeAllScreenshotsForSession)
		.flatMap(Collection::stream);

if (publishToReport) {
	screenshotStream = screenshotStream.peek(screenshot -> {
		methodContext.get().addScreenshot(screenshot);
		report.addScreenshot(screenshot, Report.FileMode.MOVE);
	});
}

final var result = screenshotStream.collect(Collectors.toList());
if (result.isEmpty()) {
	LOGGER.warn("Could not create screenshots: No sessions found in current test context.");
}
return result;
```

Reply:
But note that the warning will _also_ trigger - and claim a wrong situation - if there are simply no screenshots. That should probably be handled differently to prevent false-positives.

Reply:
Yes, the warning should appear if no sessions were found:
`methodContext.get().readSessionContexts().count() == 0`

Reply:
Then we can move it right after the first `if`, to keep these Sanity Checks in one place.